js/content.js

```diff
@@ -148,26 +106,66 @@ function pushComment(text) {
     }
 
-    //console.log(text);
-    myChatsToRender.push( {
-        isNew: true,
-        text: text,
-        time: gTextTime
-    });
+    // Ignore text if TAB is currently not active
+    if (!gTabActive)
+        return;
 
-    // To give a little bit more fluidity, keep pushing texts when
-    // tab is not active. However, each time a new chat is pushed in,
-    // make sure we update (but not render) the simulation.
-    if (!gTabActive) {
-        var currDate = new Date();
-        var elapsedSecs = (currDate.getTime() - gTabAwayTime) / 1000;
-        updateSimulation(elapsedSecs);
-        gTabAwayTime = currDate;
-        // console.log("pushed text: " + text + ", at: " + currDate);
+    // x-math
+    var canvasWidth = myTextLayer.parentElement.clientWidth;
+    var textMeasurement = myTextMeasureContext.measureText(text);
+    var textWidth = textMeasurement.width;
+    var xPos = canvasWidth;
+    var xTranslate = Math.round(canvasWidth + textWidth) + 10;
+    // To make long text scroll faster, apply a scale to them
+    var xTranslateScaled = canvasWidth + textWidth * 0.3;
+    var travelTime = getTravelTimeMs(xTranslateScaled);
+
+    // y-math
+    var LINE_HEIGHT = 40; // px - add some padding between text
+    var canvasHeight = myTextLayer.parentElement.clientHeight - LINE_HEIGHT;
+    var possibleLines = Math.floor(canvasHeight / LINE_HEIGHT) - 1;
+    // Limitting lines to the top tends to be a good thing
+    if (possibleLines > 10) {
+        possibleLines = 10;
     }
+    var randomLine = Math.round( Math.random() * possibleLines );
+    // Avoid repeating the same line twice in a row.
+    if (randomLine === gLastRandomLine) {
+        randomLine = randomLine > possibleLines/2 ? randomLine-1 : randomLine+1;
+    }
+    gLastRandomLine = randomLine;
+    var yPos = LINE_HEIGHT + (randomLine * LINE_HEIGHT); // Skip topmost line
+
+    // Create the div!
+    var sampleText = document.createElement('div');
+    sampleText.innerText = text;
+    sampleText.style.position = "absolute";
+    sampleText.style.left = xPos+"px";
+    sampleText.style.top = yPos+"px";
+    sampleText.style.color = 'white';
+    sampleText.style['white-space'] = 'nowrap';
+
+    // To get the animation to trigger, transition needs to be applied later...
+    requestAnimationFrame(function(){
+        myTextLayer.appendChild(sampleText);
+
+        // and maybe later again? I just don't know any more.
+        requestAnimationFrame(function(){
+            sampleText.style['transition'] = 'linear transform ' + travelTime + 'ms';
+            sampleText.style['transform'] = 'translateX(-' + xTranslate +'px)';
+
+            // For some reason the 'transitioned' event does not get fired.
+            // So, resolve it with a simple timeout...
+            setTimeout(function(){
+                myTextLayer.removeChild(sampleText);
+            }, travelTime);
+
+        });
+
+    });
```

Comment:
This requesetAnimationFrame chain up here is, hands down, the suckiest part of the code